network/src/noise_wrapper/handshake.rs

```diff
@@ -0,0 +1,387 @@
+                )
+            })?;
+
+        // finalize the connection
+        Ok(NoiseSession::new(socket, session))
+    }
+
+    pub async fn accept<TSocket>(
+        &self,
+        mut socket: TSocket,
+        trusted_peers: Option<&Arc<RwLock<HashMap<PeerId, NetworkPeerInfo>>>>,
+    ) -> io::Result<NoiseSession<TSocket>>
+    where
+        TSocket: AsyncRead + AsyncWrite + Unpin,
+    {
+        // receives prologue as the client timestamp in seconds
+        let mut prologue = [0u8; 8];
+        socket.read_exact(&mut prologue).await?;
+        let client_timestamp_u64 = u64::from_le_bytes(prologue);
+        let client_timestamp = time::Duration::from_secs(client_timestamp_u64);
+
+        // check the client timestamp
+        // TODO: security logging (mimoo)
+        let now = time::SystemTime::now()
+            .duration_since(time::UNIX_EPOCH)
+            .expect("system clock should work");
+        if client_timestamp > now
+            && client_timestamp - now > time::Duration::from_secs(MAX_FUTURE_TIMESTAMP)
+        {
+            println!("DEBUG: 2");
+            // if the client timestamp is too far in the future, abort
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                format!(
+                    "noise: client initiated connection with a timestamp too far in the future: {}",
+                    client_timestamp_u64
+                ),
+            ));
+        } else if now.checked_sub(client_timestamp).unwrap()
+            > time::Duration::from_secs(EXPIRATION_TIMESTAMP)
+        {
+            println!("DEBUG: 3");
+            // if the client timestamp is expired, abort
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                format!(
+                    "noise: client initiated connection with an expired timestamp: {}",
+                    client_timestamp_u64
+                ),
+            ));
+        }
+
+        // receive the initiation message
+        let mut client_init_message = [0u8; noise::handshake_init_msg_len(0)];
+        socket.read_exact(&mut client_init_message).await?;
+
+        // parse it
+        let (their_public_key, handshake_state, _) = self
+            .0
+            .parse_client_init_message(&prologue, &client_init_message)
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("noise: wrong message received {}", e),
+                )
+            })?;
+
+        // make sure the public key is a validator before continuing (if we're in the validator network)
+        if let Some(trusted_peers) = trusted_peers {
+            let found = trusted_peers
+                .read()
+                .unwrap()
+                .iter()
+                .find(|(_peer_id, public_keys)| public_keys.identity_public_key == their_public_key)
```

Comment:
seems a little interesting we don't really care if the peer knows its own identity lol, what's the point in passing this in?

Reply:
not sure I understand your question : o